src/Compilers/Core/Portable/CodeGen/DataStringHolder.cs

```diff
@@ -0,0 +1,424 @@
+using Microsoft.CodeAnalysis.Symbols;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.CodeGen;
+
+internal sealed class DataStringHolder : DefaultTypeDef, Cci.INamespaceTypeDefinition
+{
+    private const string TypeNamePrefix = "<S>";
+
+    private readonly string _name;
+    private readonly CommonPEModuleBuilder _moduleBuilder;
+    private readonly Cci.ITypeReference _systemObject;
+    private readonly Cci.ITypeReference _systemString;
+    private readonly Cci.ICustomAttribute _compilerGeneratedAttribute;
+    private readonly Cci.IMethodReference _encodingUtf8;
+    private readonly Cci.IMethodReference _encodingGetString;
+    private readonly PrivateImplementationDetails _privateImplementationDetails;
+
+    private int _frozen;
+    private ImmutableArray<SynthesizedStaticField> _orderedSynthesizedFields;
+
+    // fields mapped to metadata blocks and the corresponding `string` fields
+    private readonly ConcurrentDictionary<(ImmutableArray<byte> Data, ushort Alignment), (MappedField Bytes, StringField String)> _mappedFields =
+        new ConcurrentDictionary<(ImmutableArray<byte> Data, ushort Alignment), (MappedField Bytes, StringField String)>(PrivateImplementationDetails.DataAndUShortEqualityComparer.Instance);
+
+    // synthesized methods
+    private ImmutableArray<Cci.IMethodDefinition> _orderedSynthesizedMethods;
+
+    public DataStringHolder(
+        CommonPEModuleBuilder moduleBuilder,
+        string dataHash,
+        Cci.ITypeReference systemObject,
+        Cci.ITypeReference systemString,
+        Cci.ICustomAttribute compilerGeneratedAttribute,
+        Cci.IMethodReference encodingUtf8,
+        Cci.IMethodReference encodingGetString,
+        PrivateImplementationDetails privateImplementationDetails)
+    {
+        _name = TypeNamePrefix + dataHash;
+        _moduleBuilder = moduleBuilder;
+        _systemObject = systemObject;
+        _systemString = systemString;
+        _compilerGeneratedAttribute = compilerGeneratedAttribute;
+        _encodingUtf8 = encodingUtf8;
+        _encodingGetString = encodingGetString;
+        _privateImplementationDetails = privateImplementationDetails;
+    }
+
+    public string Name => _name;
+
+    public string NamespaceName => string.Empty;
+
+    public bool IsPublic => false;
+
+    internal bool IsFrozen => _frozen != 0;
+
+    public override Cci.INamespaceTypeDefinition AsNamespaceTypeDefinition(EmitContext context) => this;
+
+    public override Cci.INamespaceTypeReference AsNamespaceTypeReference => this;
+
+    public Cci.IFieldReference CreateDataField(ImmutableArray<byte> data)
+    {
+        Cci.ITypeReference type = _privateImplementationDetails.GetOrAddProxyType(data.Length, alignment: 1);
+
+        Debug.Assert(_mappedFields.IsEmpty, "We need to generate unique field names if we are synthesizing more than one.");
+
+        (_, StringField stringField) = _mappedFields.GetOrAdd((data, Alignment: 1), key =>
```

Comment:
>_mappedFields [](http://example.com/codeflow?start=39&length=13)

Why do we need a ```ConcurrentDictionary``` if we create at most one field?

Reply:
We could have an alternative implementation strategy that has more than one field in one class. But I will remove it for now.